### PR TITLE
[SHEPHERD] Timer api implemented

### PR DESCRIPTION
**Pull request by kylesu98**
Filled in Timer API for multithreaded Timer.

### CODE REVIEW

File: shepherd/Timer.py [hunk 1]
```diff
@@ -1,8 +1,50 @@
...
+		'''
+		When started, thread will run and process Timers in queue until manually stopped
+		TODO: Add how to send message via LCM in the case of match timer
+		'''
+		while not self.stop.isSet():
+			if(len(self.queue) > 0 and self.queue[0].endTime < time.time()):
+				queueLock.acquire()
+				event = heapq.heappop(self.queue)
+				if(event.timer_type == TIMER_TYPES.MATCH):
+					print("Send Message Via LCM to Main process") #Placeholder function
```
Comment (JHGZhou) [inline shepherd/Timer.py]: I'll show you the LCM api wednesday. Should be a quick fix

File: shepherd/Timer.py [hunk 2]
```diff
@@ -22,13 +64,47 @@ def __init__(self, timer_type, goal_name=None):
...
-        pass
+        """Starts a new timer with the duration (seconds) and sets timer to active.
+           If Timer is already running, adds duration to Timer"""
+        if(self.active):
+        	queueLock.acquire()
+        	self.endTime += time
+			heapq.heapify(Timer.eventQueue)
+			queueLock.release()
+    	else:
+    		if(!Timer.running):
```
Comment (JHGZhou) [inline shepherd/Timer.py]: not python syntax i think. `if (not Timer.running)`

File: shepherd/Timer.py [hunk 3]
```diff
@@ -1,8 +1,50 @@
...
     """
+    class busyThread(threading.Thread):
+    '''
+    Subclass that is the actual thread that will be running.
+    There will only be one for the entire class.
+    '''
+	def __init__(self, queue):
+		super(busyThread, self).__init__()
+		self.queue = queue
+		self.stop = threading.Event()
```
Comment (JHGZhou) [inline shepherd/Timer.py]: Inconsistent spacing in the file. Change it all to 4 spaces

File: shepherd/Timer.py (hunk 2 above)
Comment (brandonxxlee) [reply shepherd/Timer.py]: Don't even put the parentheses!

Comment (JHGZhou) [reply shepherd/Timer.py]: Tbh I didn't know you could leave out the parentheses for while and if in python. The more you know I guess :)

File: shepherd/Timer.py [hunk 4]
```diff
@@ -1,8 +1,50 @@
...
+		while not self.stop.isSet():
+			if(len(self.queue) > 0 and self.queue[0].endTime < time.time()):
+				queueLock.acquire()
+				event = heapq.heappop(self.queue)
+				if(event.timer_type == TIMER_TYPES.MATCH):
+					print("Send Message Via LCM to Main process") #Placeholder function
+				event.active = False
+				queueLock.release()
+
+	'''Stops this thread. Must be called from different thread (Main Thread)'''
```
Comment (JHGZhou) [inline shepherd/Timer.py]: Comment style inconsistent. 3 single quotes and under the def statement

File: shepherd/Timer.py [hunk 5]
```diff
@@ -22,13 +68,47 @@ def __init__(self, timer_type, goal_name=None):
         self.goal_name = goal_name
 
     def start_timer(self, duration):
-        """Starts a new timer with the duration (seconds) and sets timer to active"""
-        pass
+        """Starts a new timer with the duration (seconds) and sets timer to active.
```
Comment (brandonxxlee) [inline shepherd/Timer.py:71]: I think this naming could get confusing, having `start_timer` also  be used to add additional time. Perhaps split this into a separate function, or rename it?

File: shepherd/Timer.py [hunk 6]
```diff
@@ -22,13 +68,47 @@ def __init__(self, timer_type, goal_name=None):
...
+           If Timer is already running, adds duration to Timer"""
+        if(self.active):
```
Comment (brandonxxlee) [inline shepherd/Timer.py:73]: I think Python likes spaces for if statements instead of parentheses.

File: shepherd/Timer.py [hunk 7]
```diff
@@ -1,8 +1,54 @@
...
+        self.queue = queue
+        self.stop = threading.Event()
+
+    def run(self):
+        '''
+        When started, thread will run and process Timers in queue until manually stopped
+        TODO: Add how to send message via LCM in the case of match timer
+        '''
+        while not self.stop.isSet():
+            if(len(self.queue) > 0 and self.queue[0].endTime < time.time()):
```
Comment (brandonxxlee) [inline shepherd/Timer.py:29]: Small nitpick that might not ever happen, we should probably compare with `<=` instead of `<`

**Review by brandonxxlee — CHANGES_REQUESTED** [review]
overall lgtm, just minor changes

File: shepherd/Timer.py [hunk 8]
```diff
@@ -1,8 +1,54 @@
...
     and compare it to the current system time in order to see how much time is left
     """
+    class busyThread(threading.Thread):
+    '''
+    Subclass that is the actual thread that will be running.
+    There will only be one for the entire class.
+    '''
+    def __init__(self, queue):
+        super(busyThread, self).__init__()
+        self.queue = queue
```
Comment (brandonxxlee) [inline shepherd/Timer.py:20]: Actually, you probably can simplify all this synchronization logic by using the python built in `Queue.PriorityQueue`

**Review by brandonxxlee — CHANGES_REQUESTED** [review]

**Comment by brandonxxlee** [comment]
Closed, implemented in #473

**Pull request closed without merging**